kubernetes/apps_kubeclient.go

```diff
@@ -770,63 +867,433 @@ func (kc *kubeClient) getPods(namespace string, uid types.UID) ([]v1.Pod, error)
+		}
+		podStatus[statusKey]++
+		podTotal++
+	}
+
+	// if there were actually no pods, create a dummy entry
+	if podTotal == 0 {
+		podStatus[podRunning] = 0
+	}
+
+	var result [][]string
+	for status, count := range podStatus {
+		statusEntry := []string{status, strconv.Itoa(count)}
+		result = append(result, statusEntry)
+	}
+
+	return result, podTotal
+}
+
+func isPodWarning(pod *v1.Pod) (warning, severe bool) {
+	const containerTimeout time.Duration = 5 * time.Minute
+	const containerCrashLoop string = "CrashLoopBackOff"
+	// Consider Unknown phase a warning state
+	if pod.Status.Phase == v1.PodUnknown {
+		return true, false
+	}
+
+	// Check if pod has been in Pending phase for too long
+	now := time.Now()
+	if pod.Status.Phase == v1.PodPending {
+		duration := now.Sub(pod.CreationTimestamp.Time)
+		if duration > containerTimeout {
+			return true, false
+		}
+	}
+
+	// Check for warning conditions in pod's containers
+	if pod.Status.Phase == v1.PodRunning {
+		for _, status := range pod.Status.ContainerStatuses {
+			state := status.State
+			// Check if the container terminated with non-zero exit status
+			if state.Terminated != nil && state.Terminated.ExitCode != 0 {
+				// Severe if pod is terminated, indicating container didn't stop cleanly
+				return true, pod.DeletionTimestamp != nil
+			}
+			// Check if the container has been repeatedly crashing
+			if state.Waiting != nil && state.Waiting.Reason == containerCrashLoop {
+				return true, true
+			}
+			// Check if the container has not become ready within timeout
+			if state.Running != nil && !status.Ready {
+				startTime := state.Running.StartedAt.Time
+				duration := now.Sub(startTime)
+				if duration > containerTimeout {
+					return true, false
+				}
+			}
 		}
 	}
 
-	total := len(pods)
-	result := &app.PodStats{
-		Starting: &starting,
-		Running:  &running,
-		Stopping: &stopping,
-		Total:    &total,
+	return false, false
+}
+
+func isPullingImage(pod *v1.Pod) bool {
+	const containerCreating string = "ContainerCreating"
+	// If pod is pending with a container waiting due to a "ContainerCreating" event,
+	// categorize as "Pulling". This may change as more information is made available.
+	// See: https://github.com/openshift/origin-web-console/blob/v3.7.0/app/scripts/filters/resources.js#L663
+	if pod.Status.Phase == v1.PodPending {
+		for _, status := range pod.Status.ContainerStatuses {
+			waiting := status.State.Waiting
+			if waiting != nil && waiting.Reason == containerCreating {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+func isPodReady(pod *v1.Pod) bool {
+	// If all of the pod's containers have a ready status, then the pod is
+	// considered ready.
+	total := len(pod.Spec.Containers)
+	numReady := 0
+	for _, status := range pod.Status.ContainerStatuses {
+		if status.Ready {
+			numReady++
+		}
+	}
+	return numReady == total
+}
+
+func (kc *kubeClient) getBestRoute(namespace string, dc *deployment) (*url.URL, error) {
```

Comment:
is this method covered by 1 or more tests ?

Reply:
A whitebox test has not yet been written.